pkgs/development/libraries/science/math/tiny-cuda-nn/default.nix

```diff
@@ -0,0 +1,156 @@
+        pip
+        wheel
+        setuptools
+      ]);
+
+    buildInputs =
+      [
+        cuda-redist
+      ]
+      ++ lib.optionals pythonSupport (
+        with pythonPackages; [
+          python
+          pybind11
+        ]
+      );
+
+    # TODO: OpenCV, whicih also provides python bindings and supports CUDA, adds the cuda
+    #   redistributable to propagatedBuildInputs. Should we do this here as well?
+    propagatedBuildInputs = lib.optionals pythonSupport (
+      with pythonPackages; [
+        torch
+      ]
+    );
+
+    # NOTE: We cannot use pythonImportsCheck for this module because it uses torch to immediately
+    #   initailize CUDA. We cannot assume that at the time we run the check phase, the user has an
+    #   NVIDIA GPU available.
+    # There are no tests for the C++ library or the python bindings, so we just skip the check
+    # phase.
+    doCheck = false;
+
+    preConfigure = ''
+      export TCNN_CUDA_ARCHITECTURES=${
+        strings.concatStringsSep ";" (lists.map cudaFlags.dropDot cudaFlags.cudaCapabilities)
+      }
+      export CUDA_HOME=${cuda-redist}
+      export LIBRARY_PATH=${cuda-redist}/lib/stubs:$LIBRARY_PATH
+      export CC=${cudatoolkit.cc}/bin/cc
+      export CXX=${cudatoolkit.cc}/bin/c++
+      export CUDAHOSTCXX=${cudatoolkit.cc}/bin/c++
+    '';
+
+    # When building the python bindings, we cannot re-use the artifacts from the C++ build so we
+    # skip the CMake confurePhase and the buildPhase.
+    dontUseCmakeConfigure = pythonSupport;
+
+    # The configurePhase usually puts you in the build directory, so for the python bindings we
+    # need to change directories to the source directory.
+    configurePhase = strings.optionalString pythonSupport ''
+      runHook preConfigure
+      mkdir -p $NIX_BUILD_TOP/build
+      cd $NIX_BUILD_TOP/build
+      runHook postConfigure
+    '';
+
+    buildPhase = strings.optionalString pythonSupport ''
+      runHook preBuild
+      python -m pip wheel \
+        --no-build-isolation \
+        --no-clean \
+        --no-deps \
+        --no-index \
+        --verbose \
+        --wheel-dir $NIX_BUILD_TOP/build \
+        $NIX_BUILD_TOP/source/bindings/torch
+      runHook postBuild
+    '';
+
+    installPhase =
+      ''
+        runHook preInstall
+        mkdir -p $out/lib
+      ''
+      # Installing the C++ library just requires copying the static library to the output directory
+      + strings.optionalString (!pythonSupport) ''
+        cp libtiny-cuda-nn.a $out/lib/
+      ''
+      # Installing the python bindings requires building the wheel and installing it
+      + strings.optionalString pythonSupport ''
+        python -m pip install \
+          --no-build-isolation \
+          --no-cache-dir \
+          --no-deps \
+          --no-index \
+          --no-warn-script-location \
+          --prefix="$out" \
+          --verbose \
+          ./*.whl
+      ''
```

Comment:
seems like this could also be simplified by having a separate `buildPythonPackage`?

Reply:
See https://github.com/NixOS/nixpkgs/pull/218166#discussion_r1131643465